File: ios/tamerlinking/tamerlinking/Classes/LinkingModule.swift
import Foundation
import Lynx

@objcMembers
public final class LinkingModule: NSObject, LynxModule {

    @objc public static var name: String { "LinkingModule" }

    @objc public static var methodLookup: [String: String] {
        [
            "createURL": NSStringFromSelector(#selector(createURL(_:optionsJson:))),
            "getInitialURL": NSStringFromSelector(#selector(getInitialURL(_:)))
        ]
    }

    private static var pendingInitialUrl: String?
    private static weak var instance: LinkingModule?

    @objc public static func setInitialUrl(_ url: String?) {
        pendingInitialUrl = url
    }

    @objc public static func onUrlReceived(_ url: String?) {
        guard let url = url, !url.isEmpty else { return }
        instance?.emitUrl(url)
    }

    @objc public init(param: Any) {
        super.init()
        LinkingModule.instance = self
    }
    @objc public override init() {
        super.init()
        LinkingModule.instance = self
    }

    @objc func createURL(_ path: String, optionsJson: String) -> String {
        guard let data = optionsJson.data(using: .utf8),
              let opts = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return "tamerdevapp://\(path)"
        }
        let scheme = opts["scheme"] as? String ?? "tamerdevapp"
        let p = (opts["path"] as? String ?? path).trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let base = "\(scheme)://\(p)"
        if let q = opts["queryParams"] as? [String: String], !q.isEmpty {
            let pairs = q.map { "\($0.key.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? $0.key)=\($0.value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? $0.value)" }
            return "\(base)?\(pairs.joined(separator: "&"))"
        }
        return base
    }

    @objc func getInitialURL(_ callback: @escaping (String) -> Void) {
        let url = LinkingModule.pendingInitialUrl ?? ""
        if LinkingModule.pendingInitialUrl != nil {
            LinkingModule.pendingInitialUrl = nil
        }
        callback(url)
    }

    private func emitUrl(_ url: String) {
        let payload = ["url": url]
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8) else { return }
        emitEvent("tamer-linking:url", json)
    }

    private func emitEvent(_ name: String, _ payload: String) {
        DispatchQueue.main.async {
            NotificationCenter.default.post(name: NSNotification.Name(name), object: nil, userInfo: ["payload": payload])
        }
    }
}

File: android/src/main/kotlin/com/nanofuxion/tamerlinking/LinkingModule.kt
package com.nanofuxion.tamerlinking

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.react.bridge.Callback
import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.react.bridge.JavaOnlyMap
import com.lynx.tasm.behavior.LynxContext
import org.json.JSONObject

class LinkingModule(context: Context) : LynxModule(context) {

    companion object {
        private const val TAG = "LinkingModule"
        private const val URL_EVENT = "tamer-linking:url"

        @Volatile
        var instance: LinkingModule? = null
            private set

        @Volatile
        var pendingInitialUrl: String? = null

        fun setInitialUrl(url: String?) {
            pendingInitialUrl = url
        }

        fun onUrlReceived(url: String?) {
            if (url.isNullOrBlank()) return
            instance?.emitUrl(url)
        }
    }

    init {
        instance = this
    }

    @LynxMethod
    fun createURL(path: String, optionsJson: String): String {
        return try {
            val opts = JSONObject(optionsJson)
            val scheme = opts.optString("scheme", "tamerdevapp")
            val p = opts.optString("path", path).trimStart('/')
            val queryParams = opts.optJSONObject("queryParams")
            val base = "$scheme://$p"
            if (queryParams != null && queryParams.length() > 0) {
                val pairs = mutableListOf<String>()
                queryParams.keys().forEach { key ->
                    pairs.add("${Uri.encode(key)}=${Uri.encode(queryParams.getString(key))}")
                }
                "$base?${pairs.joinToString("&")}"
            } else {
                base
            }
        } catch (e: Exception) {
            Log.e(TAG, "createURL error: ${e.message}")
            "tamerdevapp://${path.trimStart('/')}"
        }
    }

    @LynxMethod
    fun getInitialURL(callback: Callback) {
        val url = pendingInitialUrl
        if (url != null) {
            pendingInitialUrl = null
        }
        callback.invoke(url ?: "")
    }

    private fun emitUrl(url: String) {
        val lynxContext = mContext as? LynxContext ?: return
        val eventDetails = JavaOnlyMap()
        eventDetails.putString("payload", JSONObject().put("url", url).toString())
        val params = JavaOnlyArray()
        params.pushMap(eventDetails)
        lynxContext.sendGlobalEvent(URL_EVENT, params)
    }
}
